api/users/client/client.go

```diff
@@ -42,3 +48,227 @@ func (c *Client) GetThread(ctx context.Context, name string) (*pb.GetThreadReply
 func (c *Client) ListThreads(ctx context.Context) (*pb.ListThreadsReply, error) {
 	return c.c.ListThreads(ctx, &pb.ListThreadsRequest{})
 }
+
+// SetupMailbox creates inbox and sentbox threads needed user mail.
+func (c *Client) SetupMail(ctx context.Context) (inbox, sentbox thread.ID, err error) {
+	res, err := c.c.SetupMail(ctx, &pb.SetupMailRequest{})
+	if err != nil {
+		return
+	}
+	inbox, err = thread.Cast(res.InboxID)
+	if err != nil {
+		return
+	}
+	sentbox, err = thread.Cast(res.SentboxID)
+	if err != nil {
+		return
+	}
+	return
+}
+
+// Message is the client side representation of a mailbox message.
+type Message struct {
+	ID        string        `json:"_id"`
+	From      thread.PubKey `json:"from"`
+	To        thread.PubKey `json:"to"`
+	Body      []byte        `json:"body"`
+	Signature []byte        `json:"signature"`
+	CreatedAt time.Time     `json:"created_at"`
+	ReadAt    time.Time     `json:"read_at,omitempty"`
+}
+
+// Open decrypts the message body with identity.
+func (m Message) Open(ctx context.Context, id thread.Identity) ([]byte, error) {
+	return id.Decrypt(ctx, m.Body)
+}
+
+// Read returns whether or not the message has been read.
+func (m Message) Read() bool {
+	return !m.ReadAt.IsZero()
+}
+
+func (m Message) Marshal() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+func (m Message) Unmarshal(data []byte) error {
+	var tm threaddb.Message
+	if err := json.Unmarshal(data, &tm); err != nil {
+		return err
+	}
+	body, err := base64.StdEncoding.DecodeString(tm.Body)
+	if err != nil {
+		return err
+	}
+	sig, err := base64.StdEncoding.DecodeString(tm.Signature)
+	if err != nil {
+		return err
+	}
+	from := &thread.Libp2pPubKey{}
+	if err := from.UnmarshalString(tm.From); err != nil {
+		return fmt.Errorf("from public key is invalid")
+	}
+	ok, err := from.Verify(body, sig)
+	if !ok || err != nil {
+		return fmt.Errorf("bad message signature")
+	}
+	to := &thread.Libp2pPubKey{}
+	if err := to.UnmarshalString(tm.To); err != nil {
+		return fmt.Errorf("to public key is invalid")
+	}
+	readAt := time.Time{}
+	if tm.ReadAt > 0 {
+		readAt = time.Unix(0, tm.ReadAt)
+	}
+	m.ID = tm.ID
+	m.From = from
+	m.To = to
+	m.Body = body
+	m.Signature = sig
+	m.CreatedAt = time.Unix(0, tm.CreatedAt)
+	m.ReadAt = readAt
+	return nil
+}
+
+// SendMessage sends the message body to a recipient.
+func (c *Client) SendMessage(ctx context.Context, from thread.Identity, to thread.PubKey, body []byte) (msg Message, err error) {
+	fromBody, err := from.GetPublic().Encrypt(body)
```

Comment:
To send a message, we create two copies of the message body:
- One encrypted for the recipient that will live in their inbox
- One encrypted for the sender that will live in their sentbox